feat(cmd): Support Base64 encoded SSH_KEY

felixbuenemann opened the pull request:
This adds a check to see if the value of SSH_KEY is already a Base64
encoded value, which is the case if "deis config:push" is used after
using "deis config:pull" on an app that already has SSH_KEY set.

This fixes #274.

felixbuenemann commented:
I tested the new behavior manually, but was unable to add automated tests, because the current location of the ssh key parsing code is poorly chosen and makes it hard to test.

In order to add sensible tests the code should probably be moved into a separate `parseSshKey` function which would be called from `parseConfig` instead of `ConfigSet`.

Joshua-Anderson commented:
We have some public ssh key parsing logic [in a package here](https://github.com/deis/workflow-cli/tree/master/pkg/ssh), maybe add the private ssh key logic parsing here too?

Also, I might swap the logic, so you check if the key is base64'd before trying to read a file and parse it, whih feels a little less hacky to me.

Thanks for contributing!

felixbuenemann commented:
@Joshua-Anderson I've thought about this and I'm wondering wether this is the right approach.

The weird thing is that `deis config` lists all environment variables in the base64-decoded form, except `SSH_KEY`. If it was properly decoding `SSH_KEY` this workaround would not be needed at all.

felixbuenemann commented:
OK, I've looked through the code and the reason that the SSH_KEY variable is base64 encoded is probably compatibility with heroku buildpack behavior. It's kinda strange, cause a private key is already in pem format which also uses base64 encoding…

felixbuenemann commented:
@Joshua-Anderson I've extracted the parsing code into a new `parseSshKey` function and added unit test for it as well as an integration style test in `ConfigSet` which ensures the code is called.

I've chosen to not put the new code into the ssh package, because it's implementation is too closely coupled to the specific behavior in ConfigSet to be re-used somewhere else, but I could move it if you disagree.

If you prefer I can squash to a single commit once the review is positive.

felixbuenemann commented:
Oh and I also changed the SSH Private Key matching regex so it also matches ed25519 or other ssh keys converted to the new openssh key format as part of a *refactor* commit. I could move that into a single commit if it's a problem.

Joshua-Anderson commented on cmd/config_test.go:
```diff
@@ -149,16 +183,17 @@ func TestConfigSet(t *testing.T) {
 	cmdr := DeisCmd{WOut: &b, ConfigFile: cf}
 
-	err = cmdr.ConfigSet("foo", []string{"TRUE=false"})
+	err = cmdr.ConfigSet("foo", []string{"TRUE=false", "SSH_KEY=-----BEGIN OPENSSH PRIVATE KEY-----"})
```
Not sure this is quite what you want, since you're checking if SSH_KEY is a base64 encoded key or a filepath.

felixbuenemann replied:
This is the integration style test, it just checks that SSH_KEY is mangled and send to the server during config:set. It's only there to ensure the `parseSshKey` code which is covered by the unit tests is called at all, which flavor of encoding (plain/path/base64) we use here doesn't really matter.

Joshua-Anderson replied:
Yeah, it just a little confusing looking over, since you're using a private key header and base64 encoding, a combo that doesn't go together in normal usage.

felixbuenemann replied:
Why doesn't it go together in normal usage?

The current docs say that you can set SSH_KEY to either a path or a private key.

The `parseSSHKey` doesn't *really* parse the key, it only checks if it *appears* to be a key by looking at the first line of the key, which is what I'm doing here.

In reality openssh wil happily use a key that just contains the base64 encoded key without any headers/footers, but that way we couldn't differentiate between a single or double base64 encoded key… There's also the golang crypto ssh module's `ParseRawPrivateKey` for actual parsing, but it only supports dss/rsa/ecdsa keys and would exclude ed25519 keys, which are perfectly fine to use here.

Maybe you could be more specific in what you would like me to change about this test.

felixbuenemann commented:
Fixed all errors from `make test-style`.

felixbuenemann commented:
@Joshua-Anderson Stil waiting on review feedback from you.

Joshua-Anderson commented:
Sorry about that. I was confused by the fact you check for ssh keys as well as ssh keys base64 encoded by deis, but I understand where you're coming from now. I'd like to manually test it, but code looks LGTM!

mboersma commented:
I've tested this change locally and it works as advertised.
